Simulation/RK4_Integration.m

```diff
@@ -1,9 +1,57 @@
 % 4th Order Runge-Kutta (RK4) integration
-function StateVector = RK4_Integration(StateVector, Object, dt)
-  k1 = dt * Derivatives(StateVector, Object);
-  k2 = dt * Derivatives(StateVector + k1/2, Object);
-  k3 = dt * Derivatives(StateVector + k2/2, Object);
-  k4 = dt * Derivatives(StateVector + k3, Object);
-
-  StateVector = StateVector + (1/6) * (k1 + 2*k2 + 2*k3 + k4);
-endfunction
\ No newline at end of file
+function SimData = RK4_Integration(SimData)
+
+    dt = SimData.dt;
+
+    % Initialize with zeroed delta
+    k0.P_delta = [0,0]; % Position Derivative
+    k0.V_delta = [0,0]; % Velocity Derivative
+    k0.O_delta = [0];   % Orientation Derivative
+    k0.Time    = SimData.Time;
+
+    % Calculate the k1 delta
+    k1_derivative = Derivatives(k0,SimData);
+    k1.P_delta = k1_derivative.P_dot*dt*0.5;
+    k1.V_delta = k1_derivative.V_dot*dt*0.5;
+    k1.O_delta = k1_derivative.O_dot*dt*0.5;
+    k1.Time    = SimData.Time + dt*0.5;
+
+    % Calculate the k2 delta
+    k2_derivative = Derivatives(k1, SimData);
+    k2.P_delta = k2_derivative.P_dot*dt*0.5;
+    k2.V_delta = k2_derivative.V_dot*dt*0.5;
+    k2.O_delta = k2_derivative.O_dot*dt*0.5;
+    k2.Time    = SimData.Time + dt*0.5;
+
+    % Calculate the k3 delta
+    k3_derivative = Derivatives(k2, SimData);
+    k3.P_delta = k3_derivative.P_dot*dt;
+    k3.V_delta = k3_derivative.V_dot*dt;
+    k3.O_delta = k3_derivative.O_dot*dt;
+    k3.Time    = SimData.Time + dt;
+
+    % Calculate the k4 delta
+    k4_derivative = Derivatives(k3, SimData);
+    k4.P_delta = k4_derivative.P_dot*dt;
+    k4.V_delta = k4_derivative.V_dot*dt;
+    k4.O_delta = k4_derivative.O_dot*dt;
```

Comment:
There is a lot of basically duplicated code here. Could we condense this into a function? Something that would take `kN` and `SimData` and return `kN+1`?

Reply:
I cut out a ton of the excess code by putting it inside of the Derivatives function. Unfortunately since each kN is not treated the same in terms of their weights I had to add an extra parameter but it allows for reducing it to more readable and nicer code.